bug fix: Large number handling error in conv_string function(#2930)

silverbullet233 opened the pull request:
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #2930;

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->
I  found some problems with the current implementation of `conv_int/conv_string` function:
1. Incorrect handling of negative src_base in both `conv_int` and `conv_string`:  if the number is greater than 0 and src_base is nagative, it will return NULL directly which is expected to be an integer. For example, `conv(10, -2, 10)=NULL(but 2 is expected)`
2. Incorrent handle of large number: when the number is too large, error result will return. For example, `conv(99999999, 16, 10)=18446744073709551615(but 2576980377 is expected)`, `conv('8ac7230489e80000', 16, 10)=18446744073709551615(but 10000000000000000000 is expected)`;
3. Incompatibility with MySQL on overflow, For example, `conv('-18446744073709551616', 10, 10)=18446744073709551615(but 0 is expected)`,`conv('-18446744073709551616', 10, -10)=-1(but 0 is expected)`

In this PR, I have fixed correctness and compatibility issues to make it behave closer to MySQL.

silverbullet233 committed:
BugFix: Fix result error in conv_string function when the input value greater than 10^18(#2930)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StarRocks/starrocks?pullRequest=3039) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks?pullRequest=3039) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StarRocks/starrocks?pullRequest=3039) it.</sub>

liuyehcf commented on be/src/exprs/vectorized/math_functions.cpp:
```diff
@@ -441,20 +441,23 @@ ColumnPtr MathFunctions::conv_string(FunctionContext* context, const starrocks::
-        if (!handle_parse_result(dest_base_value, &decimal_num, parse_res)) {
+        int64_t decimal64_num = static_cast<int64_t>(decimal_num);
+        // recheck overflow
+        if (parse_res == StringParser::PARSE_SUCCESS && decimal_num > std::numeric_limits<uint64_t>::max()) {
```
Why use `std::numeric_limits<uint64_t>::max()` but not `std::numeric_limits<int64_t>::max()`

silverbullet233 replied:
According to the description of [mysql doc](https://dev.mysql.com/doc/refman/8.0/en/mathematical-functions.html#function_conv), input value can be treated as unsigned if src_base is positive number, so we can't simply judge a value in range[INT64_MAX,UINT64_MAX] as overflow.

ZiheLiu replied:
@silverbullet233 Could we judge `dest_base_value > 0 && decimal_num > std::numeric_limits<uint64_t>::max() || dest_base_value < 0 && decimal_num > std::numeric_limits<int64_t>::max()`?

ZiheLiu commented on be/src/exprs/vectorized/math_functions.cpp:
```diff
@@ -441,20 +441,23 @@ ColumnPtr MathFunctions::conv_string(FunctionContext* context, const starrocks::
-
         if (src_base_value < 0 && decimal_num >= 0) {
             result.append_null();
             continue;
```
Why this condition returns `null`? `decimal_num` is a signed number when `src_base_value` is negative, so maybe `decimal_num >= 0`  is legal?

silverbullet233 replied:
This condition seems unreasonable and will cause the results to be inconsistent with MySQL, I have fixed it @ZiheLiu

silverbullet233 committed:
1. Bugfix: fix result error in conv_int/conv_string function when the src_base is negative;
2. Make the behavior of conv function compatible with MySQL;

silverbullet233 committed:
fix code format

silverbullet233 commented:
I have updated my change, fixed more correctness and compatibility issues, you can see more details in the description, plz review again, thx~ @liuyehcf @ZiheLiu

liuyehcf commented on be/src/exprs/vectorized/math_functions.cpp:
```diff
@@ -439,22 +434,46 @@ ColumnPtr MathFunctions::conv_string(FunctionContext* context, const starrocks::
+        uint64_t decimal64_num = StringParser::string_to_int<uint64_t>(data_ptr + digit_start_offset,
+                                                                       string_value.size - digit_start_offset,
+                                                                       std::abs(src_base_value), &parse_res);
+        if (parse_res == StringParser::PARSE_SUCCESS) {
+            if (is_signed) {
+                if (negative && decimal64_num > 0ull - std::numeric_limits<int64_t>::min()) {
+                    decimal64_num = 0ull - std::numeric_limits<int64_t>::min();
+                }
+                if (!negative && decimal64_num > std::numeric_limits<int64_t>::max()) {
+                    decimal64_num = std::numeric_limits<int64_t>::max();
+                }
+            }
+        } else if (parse_res == StringParser::PARSE_FAILURE) {
```
Could this branch merge with the branch `else if (parse_res == StringParser::PARSE_OVERFLOW)` if type of decimal64_num is `int64_t`?

silverbullet233 replied:
Yes, these two branches are both dealing with the overflow problem, but merging them will not make the implementation more concise because we can't merge the overflow check condition, so I'd prefer to keep the status quo. what do you think? @liuyehcf
